src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs

```diff
@@ -62,4 +67,107 @@ private static PooledDisposer<PooledHashSet<T>> GetPooledHashSet<T>(ImmutableArr
         instance.AddRange(items);
         return disposer;
     }
+
+    private static IEnumerable<T> Prioritize<T>(IEnumerable<T> items, Func<T, bool> isPriority)
+    {
+        using var _ = ArrayBuilder<T>.GetInstance(out var normalItems);
+
+        foreach (var item in items)
+        {
+            if (isPriority(item))
+                yield return item;
+            else
+                normalItems.Add(item);
+        }
+
+        foreach (var item in normalItems)
+            yield return item;
+    }
+
+    /// <summary>
+    /// Main utility for searching across items in a solution.  The actual code to search the item should be provided in
+    /// <paramref name="callback"/>.  Each item in <paramref name="items"/> will be processed using
+    /// <code>Parallel.ForEachAsync</code>, allowing for parallel processing of the items, with a preference towards
+    /// earlier items.
+    /// </summary>
+    private static async Task PerformParallelSearchAsync<T>(
+        IEnumerable<T> items,
+        Func<T, Action<RoslynNavigateToItem>, CancellationToken, ValueTask> callback,
+        Func<ImmutableArray<RoslynNavigateToItem>, Task> onItemsFound,
+        CancellationToken cancellationToken)
+    {
+        // Use an unbounded channel to allow the writing work to write as many items as it can find without blocking.
+        // Concurrently, the reading task will grab items when available, and send them over to the host.
+        var channel = Channel.CreateUnbounded<RoslynNavigateToItem>(s_channelOptions);
+
+        await Task.WhenAll(
```

Comment:
>await Task.WhenAll( [](http://example.com/codeflow?start=8&length=19)

This pattern seems useful. Could it be made into a class to hide the channel usage and task yield/whenall details?

Reply:
have done so in the FAR pr.  will pull out of this one in that pr.